porcupine/plugins/anchors.py

```diff
@@ -22,80 +22,97 @@ def __init__(self, tab_textwidget: tkinter.Text, linenumbers: LineNumbers) -> No
+        linenumbers.bind("<<Updated>>", self.on_linenumbers_updated, add=True)
 
-    def _get_anchors(self) -> list[str]:
+    def _get_anchor_marks(self) -> list[str]:
         return [
             anchor
             for anchor in self.tab_textwidget.mark_names()
             if anchor.startswith(self.custom_anchor_prefix)
         ]
 
-    def _get_cursor_index(self) -> str:
-        return self.tab_textwidget.index("insert linestart")
+    def _get_line_number(self, index: str) -> int:
+        return int(self.tab_textwidget.index(index).split(".")[0])
 
-    def toggle_on_off(self, event: tkinter.Event[tabs.FileTab]) -> None:
-        self.prevent_duplicate_anchors()
+    def clean_and_get_anchors(self) -> dict[int, str]:
+        anchors = {}
 
-        cursor_index = self._get_cursor_index()
-        for anchor in self._get_anchors():
-            if self.tab_textwidget.index(anchor) == cursor_index:
-                self.tab_textwidget.mark_unset(anchor)
-                self.linenumbers.do_update()
-                return
+        for mark in self._get_anchor_marks():
+            anchor_row = self._get_line_number(mark)
+            if anchor_row in anchors:
+                self.tab_textwidget.mark_unset(mark)
+            else:
+                self.tab_textwidget.mark_set(mark, f"{mark} linestart")
+                anchors[anchor_row] = mark
+
+        return anchors
+
+    def add_anchor(self, lineno: int) -> str:
+        mark = self.custom_anchor_prefix + str(time.time())
         self.tab_textwidget.mark_set(
-            self.custom_anchor_prefix + str(time.time()), "insert linestart"
+            mark, f"{lineno}.0"
         )
+        return mark
+
+    # See underlines.py and langserver.py
+    def add_from_underlines(self, junk_event: object) -> None:
+        anchors = self.clean_and_get_anchors()
+        for start in self.tab_textwidget.tag_ranges("underline:diagnostics")[::2]:
+            lineno = self._get_line_number(str(start))
+            if lineno not in anchors:
+                anchors[lineno] = self.add_anchor(lineno)
+        self.linenumbers.do_update()
 
+    def toggle(self, event: tkinter.Event[tabs.FileTab]) -> None:
+        anchors = self.clean_and_get_anchors()
+        cursor_lineno = self._get_line_number("insert")
+        if cursor_lineno in anchors:
+            self.tab_textwidget.mark_unset(anchors[cursor_lineno])
+        else:
+            self.add_anchor(cursor_lineno)
         self.linenumbers.do_update()
 
     def jump_to_next(self, event: tkinter.Event[tabs.FileTab]) -> str:
-        cursor_row = self._get_cursor_index().split(".")[0]
-        anchor_list = self._get_anchors()
-        anchor_rows = [
-            int(self.tab_textwidget.index(anchorpoint).split(".")[0]) for anchorpoint in anchor_list
-        ]
+        cursor_row = self._get_line_number("insert")
+        anchors = self.clean_and_get_anchors()
```

Comment:
This function only needs the line numbers, so something like `anchor_rows = ...().keys()`